compiler/passes/src/flattening/flatten_statement.rs

```diff
@@ -93,77 +93,95 @@ impl StatementReconstructor for Flattener<'_> {
+            statements.extend(guard_statements);
         }
+
+        // We also need to guard against early returns.
+        guards.extend(self.returns.iter().flat_map(|pair| pair.0.clone()));
+
+        if guards.is_empty() {
+            return (Statement::Assert(assert), statements);
+        }
+
+        // We need to `or` the asserted expression with the guards,
+        // so extract an appropriate expression.
+        let mut expression = match assert.variant {
+            // If the assert statement is an `assert`, use the expression as is.
+            AssertVariant::Assert(expression) => expression,
+            // If the assert statement is an `assert_eq`, construct a new equality expression.
+            AssertVariant::AssertEq(left, right) => Expression::Binary(BinaryExpression {
+                left: Box::new(left),
+                op: BinaryOperation::Eq,
+                right: Box::new(right),
+                span: Default::default(),
+                id: {
+                    // Create a new node ID for the unary expression.
+                    let id = self.node_builder.next_id();
+                    // Update the type table with the type of the unary expression.
+                    self.type_table.insert(id, Type::Boolean);
+                    id
+                },
+            }),
+            // If the assert statement is an `assert_ne`, construct a new inequality expression.
+            AssertVariant::AssertNeq(left, right) => Expression::Binary(BinaryExpression {
+                left: Box::new(left),
+                op: BinaryOperation::Neq,
+                right: Box::new(right),
+                span: Default::default(),
+                id: {
+                    // Create a new node ID for the unary expression.
+                    let id = self.node_builder.next_id();
+                    // Update the type table with the type of the unary expression.
+                    self.type_table.insert(id, Type::Boolean);
+                    id
+                },
+            }),
+        };
+
+        // The assertion will be that the original assert statement is true or one of the guards is true
+        // (ie, we either didn't follow the condition chain that led to this assert, or else we took an
+        // early return).
+        for guard in guards.into_iter() {
+            expression = Expression::Binary(BinaryExpression {
+                op: BinaryOperation::Or,
+                span: Default::default(),
+                id: {
+                    // Create a new node ID for the binary expression.
+                    let id = self.node_builder.next_id();
+                    // Update the type table with the type of the binary expression.
+                    self.type_table.insert(id, Type::Boolean);
+                    id
+                },
+                left: Box::new(expression),
+                right: Box::new(guard),
+            });
+        }
+
```

Comment:
I just realized that the subexpressions created in the body of this method don't meet the SSA requirement of simple binary expressions. It's not a huge issue, because codegen handles these cases but worth fixing or putting in a TODO.

Reply:
I think it's improved but still happens. 
For example, `AssertNeq` returns `assert(id1 != id2)` whereas SSA form would require it to be `let var$1 = id1 != id2; assert(var$1);`